_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+  })
+  
+  cat(sprintf("\nEpoch %d, training: loss: %3.5f \n", epoch, mean(train_loss)))
+  
+  net$eval()
+  valid_loss <- c()
+  
+  coro::loop(for (b in valid_dl) {
+    loss <- valid_batch(b)
+    valid_loss <- c(valid_loss, loss)
+  })
+  
+  cat(sprintf("\nEpoch %d, validation: loss: %3.5f \n", epoch, mean(valid_loss)))
+}
+
+```
+
+    Epoch 1, training: loss: 0.37961 
+
+    Epoch 1, validation: loss: 1.10699 
+
+    Epoch 2, training: loss: 0.19355 
+
+    Epoch 2, validation: loss: 1.26462 
+
+    # ...
+    # ...
+
+    Epoch 49, training: loss: 0.03233 
+
+    Epoch 49, validation: loss: 0.62286 
+
+    Epoch 50, training: loss: 0.03091 
+
+    Epoch 50, validation: loss: 0.54457
+
+It's interesting to compare performances for different settings of `teacher_forcing_ratio`. With a setting of `0.5`, training loss decreases a lot more slowly; the opposite is seen with a setting of `0`. Validation loss, however, is not affected significantly.
+
+# Evaluation
+
+The code to inspect test-set forecasts is unchanged.
+
+```{r}
+net$eval()
+
+test_preds <- vector(mode = "list", length = length(test_dl))
+
+i <- 1
+
+coro::loop(for (b in test_dl) {
+  
+  input <- b$x
+  output <- net(input$to(device = device))
+  preds <- as.numeric(output)
+  
+  test_preds[[i]] <- preds
+  i <<- i + 1
+  
+})
+
+vic_elec_jan_2014 <- vic_elec %>%
+  filter(year(Date) == 2014, month(Date) == 1)
+
+test_pred1 <- test_preds[[1]]
+test_pred1 <- c(rep(NA, n_timesteps), test_pred1, rep(NA, nrow(vic_elec_jan_2014) - n_timesteps - n_forecast))
+
+test_pred2 <- test_preds[[408]]
+test_pred2 <- c(rep(NA, n_timesteps + 407), test_pred2, rep(NA, nrow(vic_elec_jan_2014) - 407 - n_timesteps - n_forecast))
+
+test_pred3 <- test_preds[[817]]
+test_pred3 <- c(rep(NA, nrow(vic_elec_jan_2014) - n_forecast), test_pred3)
+
+
+preds_ts <- vic_elec_jan_2014 %>%
+  select(Demand) %>%
+  add_column(
+    mlp_ex_1 = test_pred1 * train_sd + train_mean,
+    mlp_ex_2 = test_pred2 * train_sd + train_mean,
+    mlp_ex_3 = test_pred3 * train_sd + train_mean) %>%
+  pivot_longer(-Time) %>%
+  update_tsibble(key = name)
+
+
+preds_ts %>%
+  autoplot() +
+  scale_colour_manual(values = c("#08c5d1", "#00353f", "#ffbf66", "#d46f4d")) +
+  theme_minimal()
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-page", fig.cap="One-week-ahead predictions for January, 2014."}
+knitr::include_graphics("images/seq2seq_preds.png")
+```
+
+Comparing this to the forecast obtained from last time's RNN-MLP combo, we don't see much of a difference. Is this surprising? To me it is. If asked to speculate about the reason, I would probably say that, likely, in all of the setups we've used so far, the main carrier of information has been the final hidden state of the RNN (the single RNN in the two previous architectures, the *encoder* RNN in this one). It will be interesting to check this hypothesis in the last part of this series, where we augment the encoder-decoder architecture by *attention*.
```

Comment:
You've got five commas and a parenthetical in the sentence beginning “If asked to speculate…”, which is a sign that you should probably rework it. I'll leave that to you, but note that the first half:
> If asked to speculate about the reason, I would probably say that, likely…

is almost all setup for actual content.

Reply:
ooooh yeah ... you're right, in fact that sentence should never have made it to editor review :-)))